inst/include/Rcpp/vector/Matrix.h

```diff
@@ -175,25 +175,25 @@ class Matrix : public Vector<RTYPE, StoragePolicy>, public MatrixBase<RTYPE, tru
 
     template <typename U>
     void fill_diag__dispatch( traits::false_type, const U& u) {
-          Shield<SEXP> elem( converter_type::get( u ) ) ;
-        int n = Matrix::ncol() ;
-        int offset = n +1 ;
-        iterator it( VECTOR::begin()) ;
-        for( int i=0; i<n; i++){
-            *it = ::Rf_duplicate( elem );
-            it += offset;
+        Shield<SEXP> elem( converter_type::get( u ) );
+
+        R_xlen_t bounds = ( Matrix::nrow() < Matrix::ncol() ) ?
+                             Matrix::nrow() : Matrix::ncol();
+        
```

Comment:
I was just squinting at this (and mostly wondering if I had written it on one line) when I realized that

```
 R_xlen_t bounds = std::min( Matrix::nrow(), Matrix::ncol() );
```

is probably even easier :)    (Subjec to testing in whether `stdLLmin()` likes the R type etc pp)